### PR TITLE
[#18] [API] As a User, I can see a list of my previous keyword results

### LINKED ISSUE #18: [API] As a User, I can see a list of my previous keyword results
Build API for #8

## Why?
In order to compare different keywords & review past results.

### Additional notes:

Exposes 2x API (Keyword list and search results)

Problems to solve:
- User can retrieve past data thanks to persistence
- User can see many keywords at once
- User can see results quickly even when many (10K +) keywords are already persisted (performance & UX to address)
- User can use his phone or desktop to navigate via UI
- API expose the feature to enable third app to consume it

No Gos:
- No search feature [yet]
- No filter feature
- No multiple keywords [yet]
- No CSV File upload [yet]

## Who Benefits?
Logged Users.

### PR DESCRIPTION
**Pull request by malparty**
- #18 

## What happened

Add API endpoints to expose keyword results data:
- Include ads/links counts in `keywords#index`
- Add `keywords#show` to expose html & links for a given keyword

## Insight

- Relationship with `result_link`:
    To reduce the needed requests, I decided to use the [included](https://jsonapi.org/format/#document-top-level) element to include `result_link` elements within `keywords/:id` queries.
- Fabricators: Some updates in here to enable to seed keywords **with** `result_links`. While keeping a way to fabricate `result_link` directly (which will Fabricate related Keywords/users) for testing the result_link model. 

## Proof Of Work

[/api/v1/keywords/408](https://documenter.getpostman.com/view/16135891/TzeTKqCb#cb4d8152-b222-499f-8df6-aa8d12250714)

```json
{
    "data": {
        "id": "410",
        "type": "keyword",
        "attributes": {
            "name": "Checkers",
            "created_at": "2021-07-09T06:55:14.356Z",
            "status": "parsed",
            "ads_top_count": 1,
            "ads_page_count": 1,
            "non_ads_result_count": 16,
            "total_link_count": 28,
            "html": "<h1>Laboriosam tempore</h1><p>Nisi [...]"
        },
        "relationships": {
            "result_links": {
                "data": [
                    {
                        "id": "826",
                        "type": "result_link"
                    },
                    [...]
                    {
                        "id": "830",
                        "type": "result_link"
                    }
                ]
            }
        }
    },
    "included": [
        {
            "id": "826",
            "type": "result_link",
            "attributes": {
                "url": "http://jerde.ca",
                "link_type": "ads_top",
                "created_at": "2021-07-09T06:55:14.365Z"
            }
        },
       [...]
        {
            "id": "830",
            "type": "result_link",
            "attributes": {
                "url": "http://dicki.com",
                "link_type": "non_ads",
                "created_at": "2021-07-09T06:55:14.386Z"
            }
        }
    ]
}
```

### CODE REVIEW

**Review by ankitkalia1195 — APPROVED** [review]

File: app/serializers/keywords_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class KeywordsSerializer
```
Comment (ankitkalia1195) [inline app/serializers/keywords_serializer.rb]: @malparty Can we have only one `KeywordSerializer` and use Conditional attributes https://github.com/jsonapi-serializer/jsonapi-serializer#conditional-attributes

**Comment by malparty** [comment]
@ankitkalia1195,
both Keyword serializers are merged together using conditional attributes:
```ruby
attribute :html, if: proc { |_, params| params[:show] }

has_many :result_links, if: proc { |_, params| params[:show] }
```

I have also updated tests (as this brings logic in the serializer) to ensure the `api/v1/keywords#index` action does not include result_links nor html.

Feel free to review/comment this commit [8e6d1d9](https://github.com/malparty/google-search-ruby/pull/66/commits/8e6d1d9fd307942bf97e575aaa1765e2c06c4b89)

**Review by ankitkalia1195 — APPROVED** [review]

File: spec/requests/api/v1/keywords_controller_spec.rb [hunk 2]
```diff
@@ -83,4 +118,126 @@
       end
     end
   end
+
+  describe 'GET #show' do
+    context 'given an invalid id' do
```
Comment (junan) [inline spec/requests/api/v1/keywords_controller_spec.rb]: Lets bring happy test cases first

File: app/controllers/api/v1/application_controller.rb [hunk 3]
```diff
@@ -20,6 +20,10 @@ class ApplicationController < ActionController::API
         )
       end
 
+      rescue_from ::ActiveRecord::RecordNotFound do
```
Comment (junan) [inline app/controllers/api/v1/application_controller.rb:23]: 👍

**Review by junan — APPROVED** [review]

**Pull request merged**